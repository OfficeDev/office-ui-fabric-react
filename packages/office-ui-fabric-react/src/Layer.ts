export * from './packages/react-surfaces/Layer';
