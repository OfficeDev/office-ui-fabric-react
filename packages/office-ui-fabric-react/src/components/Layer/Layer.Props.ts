import * as React from 'react';

export interface ILayerProps extends React.HTMLProps<HTMLDivElement> {
  /** Callback for when the layer is mounted. */
  onLayerMounted?: () => void;
}

export interface ILayerRenderer {
  /**
   * Should return an element which layers can be appended to.
   */
  getContainer(): HTMLElement;
}
