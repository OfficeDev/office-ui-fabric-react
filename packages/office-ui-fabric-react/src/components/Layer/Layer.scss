
.ms-LayerHost-overlay {
  position: absolute;
  visibility: hidden;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
}

.ms-ProjectedLayer {
  visibility: visible;
}

.ms-Layer {
  position: fixed;
  z-index: 1000000;
  top: 0;
  left: 0;
  width: 100vw;
  height: 100vh;
  visibility: hidden;
}