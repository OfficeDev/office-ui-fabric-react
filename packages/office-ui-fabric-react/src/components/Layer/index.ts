export * from './Layer';
export * from './PortalLayer.base';
export * from './Layer.types';
export * from './LayerHost';
