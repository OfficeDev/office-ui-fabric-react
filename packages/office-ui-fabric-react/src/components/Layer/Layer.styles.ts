import { ILayerStyleProps, ILayerStyles } from './Layer.types';
import { ZIndexes } from '../../Styling';

export const getStyles = (
  props: ILayerStyleProps
): ILayerStyles => {
  const {
    className,
    isNotHost
  } = props;

  return ({
    root: [
      'ms-Layer',
      isNotHost && [
        'ms-Layer--fixed',
        {
          position: 'fixed',
          zIndex: ZIndexes.Layer,
          top: 0,
          left: 0,
          width: '100vw',
          height: '100vh',
          visibility: 'hidden'
        }
      ],
      className
    ],
    content: [
      'ms-Layer-content',
      {
        visibility: 'visible'
      }
    ]
  });
};
