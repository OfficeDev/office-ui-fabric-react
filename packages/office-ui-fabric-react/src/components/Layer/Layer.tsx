import * as ReactDOM from 'react-dom';
import { styled } from '../../Utilities';
import { ILayerProps, ILayerStyleProps, ILayerStyles } from './Layer.types';
import { LayerBase } from './Layer.base';
import { PortalLayerBase } from './PortalLayer.base';
import { getStyles } from './Layer.styles';

const portalSupport: boolean = !!ReactDOM.createPortal;

// TODO: is there a reason to keep LayerBase? React 16 is minbar and Customizer will be converted to React 16 context and
//        won't work with old LayerBase. (unless we're keeping a backwards compatible version of React 15 Customizer)
export const Layer = styled<ILayerProps, ILayerStyleProps, ILayerStyles>(
  portalSupport ? PortalLayerBase : LayerBase,
  getStyles,
  undefined,
  {
    scope: 'Layer',
    fields: ['hostId', 'theme', 'styles']
  }
);
