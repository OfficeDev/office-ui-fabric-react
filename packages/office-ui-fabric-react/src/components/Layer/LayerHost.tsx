import * as React from 'react';
import * as ReactDOM from 'react-dom';
import {
  BaseComponent,
  css,
  getNativeProps,
  divProperties
} from '../../Utilities';
import { ILayerProps, ILayerRenderer } from './Layer.Props';
import { ILayerHostProps } from './LayerHost.Props';
import './LayerHost.scss';

export interface ILayerHostContext {
  layerRenderer: ILayerRenderer;
}

export interface ILayerHostState {
  hasMounted: boolean;
}

/**
 * LayerHost provides a wrapper that acts as a passthrough, rendering the given children within it, but also
 * appending a div at the end, which projects all content wrapped in the Layer components within. Projecting
 * DOM to the end of the document allows for overlaying and stacking scenarios.
 *
 * Normally you do not need to interact directly with LayerHost. If you render Layers within content that isn't
 * wrapped within a LayerHost, a LayerHost will be created and appended to the end of the document body, where
 * layer content will then be projected. However in some circumstances you want Layered content to be rendered
 * in a specific place rather than document body (for example in a popup window or contained within a scrollable
 * region.) In those cases, wrap the content wihtin a LayerHost.
 *
 * @example
 * <LayerHost>
 *   <Layer>I will at the end of LayerHost.</Layer>
 *   <div>I will render normally.</div>
 * </LayerHost>
 **/
export class LayerHost extends BaseComponent<ILayerHostProps, ILayerHostState> implements ILayerRenderer {
  public static childContextTypes = {
    layerRenderer: React.PropTypes.object
  };

  private _container: HTMLElement;

  constructor(props) {
    super(props);
    this.state = {
      hasMounted: false
    };
  }

  public getContainer(): HTMLElement {
    return this._container;
  }

  public getChildContext(): ILayerHostContext {
    return {
      layerRenderer: this
    };
  }

  public componentDidMount() {
    this.setState({ hasMounted: true });
  }

  public render() {
    let divProps = getNativeProps(this.props, divProperties);

    return (
      <div { ...divProps } className={ css('ms-LayerHost', this.props.className) }>
        { this.state.hasMounted && this.props.children }
        <SelfRenderedContent>
          <div className='ms-LayerHost-overlay' ref={ this._resolveRef('_container') } />
        </SelfRenderedContent>
      </div>
    );
  }

}

export class SelfRenderedContent extends BaseComponent<{}, {}> {
  public shouldComponentUpdate() {
    return false;
  }

  public render() {
    return this.props.children as JSX.Element;
  }
}