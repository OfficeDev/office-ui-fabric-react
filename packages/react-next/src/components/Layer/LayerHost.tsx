import * as React from 'react';
import { css } from '../../Utilities';
import { ILayerHostProps } from './LayerHost.types';
import { notifyHostChanged } from './Layer.notification';

const useUnmount = (unmountFunction: () => void) => {
  const unmountRef = React.useRef(unmountFunction);
  unmountRef.current = unmountFunction;
  React.useEffect(
    () => () => {
      if (unmountRef.current) {
        unmountRef.current();
      }
    },
    [unmountFunction],
  );
};

export const LayerHost = React.memo((props: ILayerHostProps) => {
  const { id, className } = props;

  React.useEffect(() => {
    notifyHostChanged(id!);
    // eslint-disable-next-line react-hooks/exhaustive-deps -- componentDidMount
  }, []);

  useUnmount(() => {
    notifyHostChanged(id!);
  });

  return <div {...props} className={css('ms-LayerHost', className)} />;
});
