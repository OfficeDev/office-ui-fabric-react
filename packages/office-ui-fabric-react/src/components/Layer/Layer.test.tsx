/* tslint:disable:no-unused-variable */
import * as React from 'react';
import * as ReactDOM from 'react-dom';
/* tslint:enable:no-unused-variable */

import * as ReactTestUtils from 'react-addons-test-utils';
import { Layer } from './Layer';
import { LayerHost } from './LayerHost';

let { expect } = chai;

describe('Layer', () => {

  it('can pass context through', () => {

    class Child extends React.Component<{}, {}> {
      public static contextTypes = {
        foo: React.PropTypes.string.isRequired
      };

      public context: any;

      public render() {
        return (
          <div id='child'>{ this.context.foo }</div>
        );
      }
    }

    class Parent extends React.Component<{}, {}> {
      public static childContextTypes = {
        foo: React.PropTypes.string
      };

      public getChildContext() {
        return {
          foo: 'foo'
        };
      }

      public render() {
        return (
          <div id='parent'>
            <Layer>
              <Child />
            </Layer>
          </div>
        );
      }
    }

    class App extends React.Component<{}, {}> {

      public render() {
        return (
          <div id='app'>
            <LayerHost>
              <Parent />
            </LayerHost>
          </div>
        );
      }
    }

    let app = ReactTestUtils.renderIntoDocument(<App />) as React.ReactInstance;
    let appElement = ReactDOM.findDOMNode(app);
    let parentElement = appElement.querySelector('#parent');

    expect(parentElement).is.not.empty;
    expect(parentElement.ownerDocument).is.not.empty;

    let childElement = appElement.querySelector('#child');

    expect(childElement.textContent).equals('foo');
  });

});