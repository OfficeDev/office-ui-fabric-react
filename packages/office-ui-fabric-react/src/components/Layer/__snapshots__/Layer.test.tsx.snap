// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layer renders Layer correctly 1`] = `
<span
  className="ms-layer"
>
  <div
    className=
        ms-Fabric
        ms-Layer-content
        {
          -moz-osx-font-smoothing: grayscale;
          -webkit-font-smoothing: antialiased;
          color: #333333;
          font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
          font-size: 14px;
          font-weight: 400;
          visibility: visible;
        }
        & button {
          font-family: inherit;
        }
        & input {
          font-family: inherit;
        }
        & textarea {
          font-family: inherit;
        }
        button {
          margin-bottom: 0px;
          margin-left: 0px;
          margin-right: 0px;
          margin-top: 0px;
          overflow: visible;
        }
  >
    Content
  </div>
</span>
`;
